book/scheduling-and-threading.md

```diff
@@ -93,137 +44,130 @@ class Task:
             # ...
             header, body = request(script_url, url)
             self.task_runner.schedule_task(
                 Task(self.run_script, script_url, body))
 ```
 
-That's it! Now our browser will not run scripts until after `load` has completed
-and the event loop comes around again. But before continuing, let's consider
-why this change is interesting---is it anything other than just some
-refactoring?
+To run those tasks, we need to call the `run` method on our
+`TaskRunner`, which we can do in the main event loop:
+
+``` {.python expected=False}
+if __name__ == "__main__":
+    while True:
+        # ...
+        browser.tabs[browser.active_tab].task_runner.run()
+```
+
+Here I've chosen to only run tasks on the active tab. Now our browser
+will not run scripts until after `load` has completed and the event
+loop comes around again.
 
-It used to be that we had no choice but to eval scripts right away just as they
-were loaded. But now it's pretty clear that we have a lot more control over
-when to run scripts. For example, it's easy to make a change to `TaskRunner` to
-only run one script per second, or to not run them at all during page load, or
-when a tab is not the active tab. This flexibility is quite powerful, and we
-can use it without having to dive into the guts of a `Tab` or how it loads web
-pages---all we'd have to do is implement a new `TaskRunner` heuristic.
+This change is nice---pages will load a bit faster---but there's more
+to it than that. Before this change, we no choice but to run scripts
+right away just as they were loaded. But now that running scripts is a
+`Task`, the task runner controls when it runs. It could run only one
+script per second, or only when the page is active, or only if there
+isn't a higher-priority user action to respond to. A browser could
+even have multiple task runners, optimized for different use cases.
 
 Timers and setTimeout
-====================
+=====================
 
-It's often the case that when adding a task, it's not to be run right away, but
-instead at some point in the future. One example is the
-[`setTimeout`][settimeout] JavaScript API, which provides a way to run a
-function a given number of milliseconds from now. For example,
+Tasks are a natural way to support several JavaScript APIs that ask
+for a function to be run at some point in the future. For example, the
+[`setTimeout`][settimeout] JavaScript API lets you run a function some
+number of milliseconds from now. This code prints "Callback" to the
+console one minute in the future, for example:
+
+[settimeout]: https://developer.mozilla.org/en-US/docs/Web/API/setTimeout
 
 ``` {.javascript expected=false}
-function callback() {
-    console.log('Callback')
-}
+function callback() { console.log('Callback'); }
 setTimeout(callback, 1000);
 ```
-will print "Callback" to the console log one second from now.
 
-This API *could* be implemented by recording a time associated with each new
```

Comment:
I guess you didn't like this point? The aim was to motivate why we should use these fancy thread things instead of polling, which is probably what a programmer unfamiliar with threads would have attempted.

Reply:
Oops, I meant to move this to a footnote and it got lost somewhere. I've resurrected the footnote. The point is good, but I think the first description should be a "positive" one.